Create scalar transport module

lindsayad opened the pull request:
Closes #13192

lindsayad committed:
Create scalar transport module

Closes #13192

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1202126/) on 45ee431 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/22392/site)

This comment will be updated on new commits.

moosebuild commented:
Job [App documentation](https://civet.inl.gov/job/1202633/) on 9a7acd0 : invalidated by @lindsayad

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1202668/) on 9a7acd0 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>43cb5b</th>
      <th colspan=4>#22392 9a7acd</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.31%</td>
      <td>84.30%</td>
      <td>-0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>80174</td>
      <td>80171</td>
      <td>-3</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14925</td>
      <td>14928</td>
      <td>+3</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22392/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22392/coverage/framework)


## Modules coverage

Inconsistent report tags were found between the head and base reports.
This can happen when reports are missing from either the head or the base.

Inconsistent tags:
  module-scalar_transport
  module-scalar_transport-parallel
  module-scalar_transport-recover

### Combined

<table>
  <thead>
    <tr>
      <th></th>
      <th>43cb5b</th>
      <th colspan=4>#22392 9a7acd</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>75.00%</td>
      <td>75.41%</td>
      <td>+0.41%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>45</td>
      <td>46</td>
      <td>+1</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15</td>
      <td>15</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22392/coverage_diff/combined)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22392/coverage/combined)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/22392/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/22392/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/22392/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/22392/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/22392/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/22392/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/22392/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/22392/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/22392/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/22392/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/22392/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/22392/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/22392/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/22392/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/22392/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/22392/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/22392/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/22392/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/22392/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/22392/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/22392/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/22392/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/22392/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/22392/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/22392/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/22392/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/22392/coverage/xfem)

</details>



This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1202668/) on 9a7acd0 wanted to post the following:


The following coverage requirement(s) failed:
- `scalar_transport` coverage rate 48.00% is less than the required 90.0%

cticenhour committed:
Init scalar transport module SQA (#13192)

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1203288/) on 8d0e79b wanted to post the following:


The following coverage requirement(s) failed:
- `scalar_transport` coverage rate 48.00% is less than the required 90.0%

lindsayad commented:
Thanks a ton @cticenhour!

cticenhour commented:
No problem! I just now noticed I need to make one more tweak. Doesn't impact SQA docs building, but does change how it gets rendered. Will fixup my last commit real quick.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1203964/) on c0b3911 wanted to post the following:


The following coverage requirement(s) failed:
- `scalar_transport` coverage rate 48.00% is less than the required 90.0%

lindsayad committed:
Use ScalarTransportApp methods where appropriate

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1204207/) on d44a08d wanted to post the following:


The following coverage requirement(s) failed:
- `scalar_transport` coverage rate 76.00% is less than the required 90.0%

lindsayad committed:
Apply missing header

lindsayad committed:
Update BEA copyright years

lindsayad committed:
Migrate some recombination/dissociation bcs from TMAP

lindsayad committed:
Doxygen

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1205026/) on 4fb60d1 wanted to post the following:


The following coverage requirement(s) failed:
- `scalar_transport` coverage rate 89.47% is less than the required 90.0%

lindsayad committed:
Add LM objects for constraining scalar concentrations

LM multipliers can be used to enforce positivity on scalar
concentrations

loganharbour commented on modules/scalar_transport/src/kernels/TimeDerivativeLM.C:
```diff
@@ -0,0 +1,30 @@
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "TimeDerivativeLM.h"
+#include "Function.h"
```
Unnecessary include I think

lindsayad commented on modules/scalar_transport/doc/moosedocs.py:
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env python3
```
I did

lindsayad commented on modules/scalar_transport/include/kernels/BodyForceLM.h:
```diff
@@ -0,0 +1,31 @@
+/************************************************************/
+/*                DO NOT MODIFY THIS HEADER                 */
+/*   TMAP8: Tritium Migration Analysis Program, Version 8   */
+/*                                                          */
+/*   Copyright 2021 - 2022 Battelle Energy Alliance, LLC    */
+/*                   ALL RIGHTS RESERVED                    */
+/************************************************************/
```
Ugh I got the source files ... how did I miss the headers!

lindsayad committed:
Address Logan review

- fix header headings
- switch name of `v` variable in LMDiffusion to `primal_variable`
- use `paramError` instead of `mooseError`
- change `moosedocs.py` to symlink

lindsayad commented:
@cticenhour I need to add a dependency for the `misc` module for documentation. I tried to follow what is done in `thermal_hydraulics` but that lead me to errors like
```
MooseDocs.base.readers (ForkProcess-15): 
TOKENIZE ERROR: '/Materials/WallHeatTransferCoefficient3EqnDittusBoelterMaterial' syntax was not recognized.
/home/lindad/projects/moose/modules/thermal_hydraulics/doc/content/modules/thermal_hydraulics/source/materials/WallHeatTransferCoefficient3EqnDittusBoelterMaterial.md:12
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
12│                                                                                                │
13│                                                                                                │
14│!syntax inputs /Materials/WallHeatTransferCoefficient3EqnDittusBoelterMaterial                  │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
```

GiudGiud commented:
we ll need the TMAP issues for these objects moved to MOOSE?

cticenhour committed:
Register dependent module syntax with ScalarTransportApp and add documentation content to config.yml

cticenhour committed:
Add scalar transport module landing page

lindsayad commented:
> we ll need the TMAP issues for these objects moved to MOOSE?

I would need an SQA expert to tell me if that's necessary. Unlike for THM the TMAP8 repository is not being archived. It's issue tracker is still active

cticenhour commented:
> Unlike for THM the TMAP8 repository is not being archived. It's issue tracker is still active

A couple things here. There is the capability in MooseDocs to refer to another repo's issues for tracking purposes. We would need to make a small change to the `moose/modules/doc/config.yml` to allow for `TMAP8#1234` in an `issues` parameter for a test spec. 

That establishes what we *can* do. Now whether that's something we *want* to do is another question. My opinion is that TMAP8 would likely then need to be subject to some measure of INL-based backup and archival alongside MOOSE if we want to start referring to TMAP8 issues (even if we never do it again beyond this point). We can't lose linked issues in our requirements listing if something catastrophic were to happen and we lose things on GitHub. To my knowledge, MOOSE is the only public repo being actively backed up along with its issues (because we're required to do so). 

So, I would say we do need to duplicate some issues (with GitHub links to the original TMAP8 issues for reference) in MOOSE to reflect the prompting issue (and containing all relevant context) for each of the tests/requirements being added. @permcody might also want to chime in with his own opinion.

lindsayad commented:
Can we selectively apply whatever issue migration script was used for thermal hydraulics?

loganharbour commented:
> Can we selectively apply whatever issue migration script was used for thermal hydraulics? Only would currently need to apply it for [idaholab/TMAP8#13](https://github.com/idaholab/TMAP8/issues/13) and [idaholab/TMAP8#14](https://github.com/idaholab/TMAP8/issues/14)

Yeah, we can. I'm the one that did the transfer and still have it available. Let me know

lindsayad commented:
Yea let's do it. Just those two issues, at this time at least

lindsayad committed:
Correct issues references for tests

loganharbour commented on modules/modules.mk:
```diff
@@ -94,8 +95,20 @@ ifeq ($(XFEM),yes)
 endif
 
+ifeq ($(SCALAR_TRANSPORT),yes)
+	CHEMICAL_REACTIONS          := yes
+	NAVIER_STOKES               := yes
+	THERMAL_HYDRAULICS	    := yes
+        FLUID_PROPERTIES            := yes
+        HEAT_CONDUCTION             := yes
+        RDG                         := yes
+        RAY_TRACING                 := yes
+        SOLID_PROPERTIES            := yes
+        MISC                        := yes
```
```suggestion
	      CHEMICAL_REACTIONS          := yes
	      NAVIER_STOKES               := yes
	      THERMAL_HYDRAULICS	        := yes
        FLUID_PROPERTIES            := yes
        HEAT_CONDUCTION             := yes
        RDG                         := yes
        RAY_TRACING                 := yes
        SOLID_PROPERTIES            := yes
        MISC                        := yes
```

cticenhour replied:
What's with all the extra spaces here?

cticenhour replied:
I see now - weird how the rendering is different in the snippet view compared to the overview.

loganharbour commented on modules/scalar_transport/doc/content/source/bcs/BinaryRecombinationBC.md:
```diff
@@ -0,0 +1,22 @@
+\int_{\Omega} \psi_i K_r uv d\Omega
+\end{equation}
+
+where $K_r$ is the recombination coefficient specified with the `Kr` parameter,
```
```suggestion
where $K_r$ is the recombination coefficient specified with [!param](/BCs/BinaryRecombinationBC/Kr),
```

loganharbour commented on modules/scalar_transport/doc/content/source/bcs/BinaryRecombinationBC.md:
```diff
@@ -0,0 +1,22 @@
+\end{equation}
+
+where $K_r$ is the recombination coefficient specified with the `Kr` parameter,
+$u$ corresponds to `variable` and $v$ is a coupled variable specified by the `v`
+parameter. As the name states, this class is meant to model binary recombination
```
```suggestion
$u$ corresponds to [!param](/BCs/BinaryRecombinationBC/variable) and $v$ is a coupled variable specified by [!param](/BCs/BinaryRecombinationBC/v).
As the name states, this class is meant to model binary recombination
```

loganharbour commented on modules/scalar_transport/doc/content/source/bcs/DissociationFluxBC.md:
```diff
@@ -0,0 +1,22 @@
+- \int_{\Omega} \psi_i K_d v d\Omega
+\end{equation}
+
+where $K_d$ is a dissociation coefficient specified with the `Kd` parameter, and
+$v$ is the coupled variable specified with the `v` parameter that is
+dissociating at the boundary to create the species corresponding to the
+`variable` parameter.
```
```suggestion
where $K_d$ is a dissociation coefficient specified with [!param](/BCs/DissociationFluxBC/Kd) and
$v$ is the coupled variable specified with [!param](/BCs/DissociationFluxBC/v) that is
dissociating at the boundary to create the species corresponding to
[!param](/BCs/DissociationFluxBC/variable).
```

loganharbour commented on modules/scalar_transport/doc/content/source/bcs/UnaryRecombinationBC.md:
```diff
@@ -0,0 +1,20 @@
+\int_{\Omega} \psi_i K_r u^2 d\Omega
+\end{equation}
+
+where $K_r$ is the recombination coefficient specified with the `Kr` parameter and
+$u$ corresponds to the `variable` parameter.
```
```suggestion
where $K_r$ is the recombination coefficient specified with [!param](/BCs/UnaryRecombinationBC/Kr)
and $u$ corresponds to [!param](/BCs/UnaryRecombinationBC/variable).
```

loganharbour reviewed (changes requested):
Good to go after this; you should be able to batch these changes

GiudGiud commented on modules/modules.mk:
```diff
@@ -34,6 +34,7 @@ ifeq ($(ALL_MODULES),yes)
         RDG                         := yes
         REACTOR                     := yes
         RICHARDS                    := yes
+	SCALAR_TRANSPORT            := yes
```
same for spaces here

GiudGiud commented on modules/modules.mk:
```diff
@@ -94,8 +95,20 @@ ifeq ($(XFEM),yes)
         TENSOR_MECHANICS            := yes
 endif
 
+ifeq ($(SCALAR_TRANSPORT),yes)
```
and here

GiudGiud commented on modules/scalar_transport/include/bcs/BinaryRecombinationBC.h:
```diff
@@ -0,0 +1,34 @@
+protected:
+  ADReal computeQpResidual() override;
+
+  /// The concentration of B, e.g. the concentration of the other specie recombining with the
```
```suggestion
  /// The concentration of B, e.g. the concentration of the other species recombining with the
```

GiudGiud commented on modules/scalar_transport/include/kernels/CoupledForceLM.h:
```diff
@@ -0,0 +1,27 @@
+
+#include "LMKernel.h"
+
+class CoupledForceLM : public LMKernel
```
docstring

GiudGiud commented on modules/scalar_transport/include/kernels/LMKernel.h:
```diff
@@ -0,0 +1,35 @@
+
+#include "ADKernelValue.h"
+
+class LMKernel : public ADKernelValue
```
docstrings

GiudGiud commented on modules/scalar_transport/include/kernels/LMTimeKernel.h:
```diff
@@ -0,0 +1,23 @@
+
+#pragma once
+
+#include "LMKernel.h"
```
docstrings

GiudGiud commented on modules/scalar_transport/src/bcs/BinaryRecombinationBC.C:
```diff
@@ -0,0 +1,34 @@
+BinaryRecombinationBC::validParams()
+{
+  auto params = ADIntegratedBC::validParams();
+  params.addClassDescription("Models recombination of the variable with a coupled specie at "
```
```suggestion
  params.addClassDescription("Models recombination of the variable with a coupled species at "
```

GiudGiud commented on modules/scalar_transport/src/bcs/UnaryRecombinationBC.C:
```diff
@@ -0,0 +1,33 @@
+InputParameters
+UnaryRecombinationBC::validParams()
+{
+  auto params = ADIntegratedBC::validParams();
```
do we really need that? it s the same as binary with the same parameter

lindsayad replied:
@loganharbour what do you mean by the 👎 ?

loganharbour replied:
I mean that I see the difference and am fine with the two objects

GiudGiud commented on modules/scalar_transport/test/tests/ncp-lms/README.md:
```diff
@@ -0,0 +1,200 @@
+# Summary of constraint enforcement techniques
```
this file should be used in some capacity. looks like it's just there? is it linked in the docs?

loganharbour replied:
It's a detailed description of the tests in the form of a markdown, not in official docs. this is fine. it can be used as a starting point for detailed doco in the future.

GiudGiud commented on modules/scalar_transport/test/tests/ncp-lms/elemental.cmp:
```diff
@@ -0,0 +1,9 @@
+COORDINATES absolute 1.e-6
+
+TIME STEPS relative 1.e-6 floor 0.0
+
```
not used?

lindsayad committed:
Apply suggestions from code review

- use param feature of moose docs
- whitespace fixup; tabs to spaces

Co-authored-by: Logan Harbour <loganhharbour@gmail.com>

loganharbour commented:
In comes the triple review with @GiudGiud... sorry @lindsayad 

ping me before merging and I'll get civet stuff ready

lindsayad committed:
Address Guillaume review

- Add/improve doxygen
- Remove kernel that shouldn't have been there  in nodal force
  application test
- Remove documentation of NCP tests for inputs that have been
  removed

lindsayad committed:
Remove UnaryRecombinationBC

lindsayad commented:
@loganharbour looks like we're about ready
